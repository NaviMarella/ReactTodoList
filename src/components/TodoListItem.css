.listitemContainer{
    display: flex;
    height: 100%;
    max-height: 80px;
    align-items: center;
    padding: 20px;
    border: 1px solid lightgray;
}

.listitemContainer .label{
    text-align: left;
    font-size: 20px;
}
