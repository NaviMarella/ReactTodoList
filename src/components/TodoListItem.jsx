import PropTypes from 'prop-types';
import React, { Component } from "react";
import deleteIcon from "../icons/delete.svg";
import editIcon from "../icons/edit.svg";
import tickIcon from "../icons/tick.svg";
import Checkbox from './Checkbox';
import Conditional from './Conditional';
import Icon from "./Icon";
import Input from "./Input";
import Label from "./Label";
import "./TodoListItem.css";

class TodoListItem extends Component{
    constructor(props){
        super(props);
        this.state={
            isEditMode: false,
            currentText: ""
        }
    }


    handleCheckboxChange = () => {
        const{ dispatch, todo } = this.props;
        dispatch({
            type: 'TOGGLE_COMPLETE',
            payload: {
                id: todo.id
            }
        })
    }

    handleInputChange = (val) => {
        this.setState({ currentText: val });
    }

    handleEditClick = () => {
        const { todo } = this.props;
        this.setState({
            isEditMode: true,
            currentText:todo.text
        })
    }

    handleDeleteClick = () => {
       const{ dispatch, todo } = this.props;
        dispatch({
            type: 'DELETE_TODO',
            payload: { id: todo.id }
        })
    }

    handleTickClick = () => {
        this.setState({
            isEditMode: false
        })
        const{ dispatch, todo } = this.props;
        dispatch({
            type: 'EDIT_TODO',
            payload: {
                id: todo.id,
                text: this.state.currentText
            }
        })
    }

    render(){
        const {todo = {}} = this.props;
        const { isEditMode, currentText } = this.state;
        const {text, isCompleted, id} = todo;
        return (
            <div className="listitemContainer">
                <Checkbox id={id} handleCheckboxChange ={this.handleCheckboxChange} isChecked={isCompleted}/>
                <Conditional condition={!isEditMode}>
                    <Label text={text} crossText={isCompleted}/>
                    <Icon icon={editIcon} onClick={this.handleEditClick}/>
                </Conditional>
                <Conditional condition={isEditMode}>
                    <Input type="text" value={currentText}  ref={this.exampleRef} handleInput={this.handleInputChange}/>
                    <Icon icon={tickIcon} onClick={this.handleTickClick}/>
                </Conditional>
                <Icon icon={deleteIcon} onClick={this.handleDeleteClick} />
            </div>
        )
    }
}

TodoListItem.propTypes = {
    todo: PropTypes.shape({}).isRequired,
    dispatch: PropTypes.func.isRequired
};

export default TodoListItem;