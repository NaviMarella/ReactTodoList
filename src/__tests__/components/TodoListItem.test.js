import { shallow } from 'enzyme';
import React from 'react';
import TodoListItem from './../../components/TodoListItem';

describe('TodoListItem Component', () => {
  test('render snapshot', () => {
      let props = {
        todo: {
            id: 1,
            label: 'TODOS',
            text: "Learn React",
            isCompleted: true
        },
        dispatch: jest.fn()
      }
    const wrapper = shallow(<TodoListItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('render snapshot for events', () => {
      let props = {
        todo: {
            id: 1,
            label: 'TODOS',
            text: "Learn React",
            isCompleted: true
        },
        dispatch: jest.fn()
      }
    const wrapper = shallow(<TodoListItem {...props} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.instance().handleCheckboxChange();
    wrapper.instance().handleDeleteClick();
    expect(props.dispatch).toHaveBeenCalled();
  });

  test('render snapshot for tick events', () => {
      let props = {
        todo: {
            id: 1,
            label: 'TODOS',
            text: "Learn React",
            isCompleted: true
        },
        dispatch: jest.fn()
      }
    const wrapper = shallow(<TodoListItem {...props} />);
    wrapper.instance().handleEditClick();
    expect(wrapper.state('isEditMode')).toEqual(true);

    wrapper.instance().handleTickClick();
    expect(wrapper.state("isEditMode")).toEqual(false);
  });

  test('render snapshot for delete events', () => {
      let props = {
        todo: {
            id: 1,
            label: 'TODOS',
            text: "Learn React",
            isCompleted: true
        },
        dispatch: jest.fn()
      }
    const wrapper = shallow(<TodoListItem {...props} />);
     wrapper.setState({currentText: ""})
    wrapper.instance().handleInputChange("Test");
    expect(wrapper.state('currentText')).toEqual("Test");
  });
});
