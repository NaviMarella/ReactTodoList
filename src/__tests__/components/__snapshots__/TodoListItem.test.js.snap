// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem Component render snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoListItem
    dispatch={[MockFunction]}
    todo={
      Object {
        "id": 1,
        "isCompleted": true,
        "label": "TODOS",
        "text": "Learn React",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Checkbox
          handleCheckboxChange={[Function]}
          id={1}
          isChecked={true}
          type="checkbox"
        />,
        <Conditional
          condition={true}
        >
          <Label
            crossText={true}
            text="Learn React"
          />
          <Icon
            icon="edit.svg"
            onClick={[Function]}
          />
        </Conditional>,
        <Conditional
          condition={false}
        >
          <Input
            handleInput={[Function]}
            maxLength={50}
            placeholder="Take a Note"
            type="text"
            value=""
          />
          <Icon
            icon="tick.svg"
            onClick={[Function]}
          />
        </Conditional>,
        <Icon
          icon="delete.svg"
          onClick={[Function]}
        />,
      ],
      "className": "listitemContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "handleCheckboxChange": [Function],
          "id": 1,
          "isChecked": true,
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Label
              crossText={true}
              text="Learn React"
            />,
            <Icon
              icon="edit.svg"
              onClick={[Function]}
            />,
          ],
          "condition": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "crossText": true,
              "text": "Learn React",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "icon": "edit.svg",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Input
              handleInput={[Function]}
              maxLength={50}
              placeholder="Take a Note"
              type="text"
              value=""
            />,
            <Icon
              icon="tick.svg"
              onClick={[Function]}
            />,
          ],
          "condition": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleInput": [Function],
              "maxLength": 50,
              "placeholder": "Take a Note",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "icon": "tick.svg",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "icon": "delete.svg",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Checkbox
            handleCheckboxChange={[Function]}
            id={1}
            isChecked={true}
            type="checkbox"
          />,
          <Conditional
            condition={true}
          >
            <Label
              crossText={true}
              text="Learn React"
            />
            <Icon
              icon="edit.svg"
              onClick={[Function]}
            />
          </Conditional>,
          <Conditional
            condition={false}
          >
            <Input
              handleInput={[Function]}
              maxLength={50}
              placeholder="Take a Note"
              type="text"
              value=""
            />
            <Icon
              icon="tick.svg"
              onClick={[Function]}
            />
          </Conditional>,
          <Icon
            icon="delete.svg"
            onClick={[Function]}
          />,
        ],
        "className": "listitemContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "handleCheckboxChange": [Function],
            "id": 1,
            "isChecked": true,
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Label
                crossText={true}
                text="Learn React"
              />,
              <Icon
                icon="edit.svg"
                onClick={[Function]}
              />,
            ],
            "condition": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "crossText": true,
                "text": "Learn React",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "icon": "edit.svg",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Input
                handleInput={[Function]}
                maxLength={50}
                placeholder="Take a Note"
                type="text"
                value=""
              />,
              <Icon
                icon="tick.svg"
                onClick={[Function]}
              />,
            ],
            "condition": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleInput": [Function],
                "maxLength": 50,
                "placeholder": "Take a Note",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "icon": "tick.svg",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "icon": "delete.svg",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`TodoListItem Component render snapshot for events 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoListItem
    dispatch={[MockFunction]}
    todo={
      Object {
        "id": 1,
        "isCompleted": true,
        "label": "TODOS",
        "text": "Learn React",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Checkbox
          handleCheckboxChange={[Function]}
          id={1}
          isChecked={true}
          type="checkbox"
        />,
        <Conditional
          condition={true}
        >
          <Label
            crossText={true}
            text="Learn React"
          />
          <Icon
            icon="edit.svg"
            onClick={[Function]}
          />
        </Conditional>,
        <Conditional
          condition={false}
        >
          <Input
            handleInput={[Function]}
            maxLength={50}
            placeholder="Take a Note"
            type="text"
            value=""
          />
          <Icon
            icon="tick.svg"
            onClick={[Function]}
          />
        </Conditional>,
        <Icon
          icon="delete.svg"
          onClick={[Function]}
        />,
      ],
      "className": "listitemContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "handleCheckboxChange": [Function],
          "id": 1,
          "isChecked": true,
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Label
              crossText={true}
              text="Learn React"
            />,
            <Icon
              icon="edit.svg"
              onClick={[Function]}
            />,
          ],
          "condition": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "crossText": true,
              "text": "Learn React",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "icon": "edit.svg",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Input
              handleInput={[Function]}
              maxLength={50}
              placeholder="Take a Note"
              type="text"
              value=""
            />,
            <Icon
              icon="tick.svg"
              onClick={[Function]}
            />,
          ],
          "condition": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleInput": [Function],
              "maxLength": 50,
              "placeholder": "Take a Note",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "icon": "tick.svg",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "icon": "delete.svg",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Checkbox
            handleCheckboxChange={[Function]}
            id={1}
            isChecked={true}
            type="checkbox"
          />,
          <Conditional
            condition={true}
          >
            <Label
              crossText={true}
              text="Learn React"
            />
            <Icon
              icon="edit.svg"
              onClick={[Function]}
            />
          </Conditional>,
          <Conditional
            condition={false}
          >
            <Input
              handleInput={[Function]}
              maxLength={50}
              placeholder="Take a Note"
              type="text"
              value=""
            />
            <Icon
              icon="tick.svg"
              onClick={[Function]}
            />
          </Conditional>,
          <Icon
            icon="delete.svg"
            onClick={[Function]}
          />,
        ],
        "className": "listitemContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "handleCheckboxChange": [Function],
            "id": 1,
            "isChecked": true,
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Label
                crossText={true}
                text="Learn React"
              />,
              <Icon
                icon="edit.svg"
                onClick={[Function]}
              />,
            ],
            "condition": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "crossText": true,
                "text": "Learn React",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "icon": "edit.svg",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Input
                handleInput={[Function]}
                maxLength={50}
                placeholder="Take a Note"
                type="text"
                value=""
              />,
              <Icon
                icon="tick.svg"
                onClick={[Function]}
              />,
            ],
            "condition": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleInput": [Function],
                "maxLength": 50,
                "placeholder": "Take a Note",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "icon": "tick.svg",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "icon": "delete.svg",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
